runtime/io/io.cpp

```diff
@@ -0,0 +1,440 @@
+    case EADDRINUSE: errStr = ERRTAG(EADDRINUSE); break;
+    case EADDRNOTAVAIL: errStr = ERRTAG(EADDRNOTAVAIL); break;
+    case ENETDOWN: errStr = ERRTAG(ENETDOWN); break;
+    case ENETUNREACH: errStr = ERRTAG(ENETUNREACH); break;
+    case ENETRESET: errStr = ERRTAG(ENETRESET); break;
+    case ECONNABORTED: errStr = ERRTAG(ECONNABORTED); break;
+    case ECONNRESET: errStr = ERRTAG(ECONNRESET); break;
+    case ENOBUFS: errStr = ERRTAG(ENOBUFS); break;
+    case EISCONN: errStr = ERRTAG(EISCONN); break;
+    case ENOTCONN: errStr = ERRTAG(ENOTCONN); break;
+    case ESHUTDOWN: errStr = ERRTAG(ESHUTDOWN); break;
+    case ETOOMANYREFS: errStr = ERRTAG(ETOOMANYREFS); break;
+    case ETIMEDOUT: errStr = ERRTAG(ETIMEDOUT); break;
+    case ECONNREFUSED: errStr = ERRTAG(ECONNREFUSED); break;
+    case EHOSTDOWN: errStr = ERRTAG(EHOSTDOWN); break;
+    case EHOSTUNREACH: errStr = ERRTAG(EHOSTUNREACH); break;
+    case ELOOP: errStr = ERRTAG(ELOOP); break;
+    case EOVERFLOW: errStr = ERRTAG(EOVERFLOW); break;
+    default:
+      block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(uint64_t)));
+      retBlock->h = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("Lbl'Hash'unknownIOError"));
+      mpz_t err;
+      mpz_init_set_si(err, errno);
+      mpz_ptr p = move_int(err);
+      memcpy(retBlock->children, &p, sizeof(mpz_ptr));
+      return retBlock;
+    }
+
+    return (block *)((((uint64_t)getTagForSymbolName(errStr)) << 32) | 1);
+  }
+
+  blockheader header_int() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      header = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortInt, SortIOInt}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_err() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      header = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortIOError, SortKItem}"));
+    }
+
+    return header;
+  }
+
+  blockheader header_string() {
+    static blockheader header = {(uint64_t)-1};
+
+    if (header.hdr == -1) {
+      header = getBlockHeaderForSymbol((uint64_t)getTagForSymbolName("inj{SortString, SortIOString}"));
+    }
+
+    return header;
+  }
+
+  inline block * getKSeqErrorBlock() {
+    block * err = block_errno();
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + 2 * sizeof(block *)));
+    block * inj = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(block *)));
+    retBlock->h = kseqHeader;
+    inj->h = header_err();
+    memcpy(inj->children, &err, sizeof(block *));
+    memcpy(retBlock->children, &inj, sizeof(block *));
+    memcpy(&(retBlock->children[1]), &dotK, sizeof(block *));
+    return retBlock;
+  }
+
+  inline block* getInjErrorBlock() {
+    block * retBlock = static_cast<block *>(koreAlloc(sizeof(block) + sizeof(block *)));
+    retBlock->h = header_err();
+    block * p = block_errno();
```

Comment:
it's probably incorrect to put this line after the alloc since it will screw up the errno.